mlir_graphblas/src/lib/GraphBLAS/GraphBLASLowerPass.cpp

```diff
@@ -779,57 +779,117 @@ class LowerMatrixReduceToVectorRewrite
           {
-            rewriter.setInsertionPointToStart(rowSumLoop.getBody());
-            Value currentSum = rowSumLoop.getLoopBody().getArgument(1);
-            Value currentPtr = rowSumLoop.getInductionVar();
-            Value rowValue =
-                rewriter.create<memref::LoadOp>(loc, matrixValues, currentPtr);
-            Value updatedSum =
-                llvm::TypeSwitch<Type, Value>(elementType)
+            Value ptr = rewriter.create<IndexCastOp>(loc, ptr64, indexType);
+            Value nextPtr =
+                rewriter.create<IndexCastOp>(loc, nextPtr64, indexType);
+            scf::ForOp rowSumLoop = rewriter.create<scf::ForOp>(
+                loc, ptr, nextPtr, c1, ValueRange{c0_elementType});
+            {
+              rewriter.setInsertionPointToStart(rowSumLoop.getBody());
+              Value currentSum = rowSumLoop.getLoopBody().getArgument(1);
+              Value currentPtr = rowSumLoop.getInductionVar();
+              Value rowValue = rewriter.create<memref::LoadOp>(
+                  loc, matrixValues, currentPtr);
+              Value updatedSum =
+                  llvm::TypeSwitch<Type, Value>(elementType)
+                      .Case<IntegerType>([&](IntegerType type) {
+                        return rewriter
+                            .create<AddIOp>(loc, rowValue, currentSum)
+                            .getResult();
+                      })
+                      .Case<FloatType>([&](FloatType type) {
+                        return rewriter
+                            .create<AddFOp>(loc, rowValue, currentSum)
+                            .getResult();
+                      });
+              rewriter.create<scf::YieldOp>(loc, ValueRange{updatedSum});
+              rewriter.setInsertionPointAfter(rowSumLoop);
+            }
+            Value rowSum = rowSumLoop.getResult(0);
+            rewriter.create<memref::StoreOp>(loc, rowSum, outputValues,
+                                             outputValuesPosition);
+            Value rowIndex64 =
+                rewriter.create<IndexCastOp>(loc, rowIndex, int64Type);
+            rewriter.create<memref::StoreOp>(loc, rowIndex64, outputIndices,
+                                             outputValuesPosition);
+            Value updatedOutputValuesPosition =
+                rewriter.create<AddIOp>(loc, outputValuesPosition, c1)
+                    .getResult();
+            rewriter.create<scf::YieldOp>(
+                loc, ValueRange{updatedOutputValuesPosition});
+          }
+
+          rewriter.setInsertionPointToStart(ifRowIsNonEmptyBlock.elseBlock());
+          {
+            rewriter.create<scf::YieldOp>(loc,
+                                          ValueRange{outputValuesPosition});
+          }
+
+          rewriter.setInsertionPointAfter(ifRowIsNonEmptyBlock);
+        }
+      } else if (aggregator == "count") {
+
+        Value ptrDiff = rewriter.create<SubIOp>(loc, nextPtr64, ptr64);
+
```

Comment:
I understand why you wrote all this code -- to compute the count in the most efficient way possible.
However, I'm not sure we actually need to do this. In theory, if we write suboptimal code, treating "count" like any other reducer, other than that it doesn't depend on the input values, the MLIR and LLVM optimization passes should recognize this fact and optimize the loop away.

I'm okay leaving this code for now, but we should have a discussion about how much we specialize code for each operator when there is a fast path vs how much we trust MLIR and LLVM to optimize things automatically.

Reply:
This discussion sounds like a good idea. 

I wasn't 100% sure that there was some MLIR/LLVM pass that optimizes the loop away, so I added in the optimization. If there is one, then we should add a code comment stating that near the code for `count`. 

I was writing this code using the heuristic of "optimize at a higher level dialect as much as possible".